handshake/src/enc/encoder.rs

```diff
@@ -0,0 +1,285 @@
+    pub fn as_string(&self) -> String {
+        match &self {
+            Address::WPKH(s) => s.clone(),
+            Address::WSH(s) => s.clone(),
+            Address::OP_RETURN(s) => s.clone(),
+        }
+    }
+
+    /// Convert the address to an `addr()` descriptor
+    pub fn to_descriptor(&self) -> String {
+        format!("addr({})", self.as_string())
+    }
+}
+
+/// NetworkParams holds the encoding paramteres for a bitcoin-like network. Currently this is
+/// composed of the bech32 human-readable prefix for witness addresses.
+pub trait NetworkParams {
+    /// The BECH32 HRP. "hs" for mainnet.
+    const HRP: &'static str;
+}
+
+/// Marker trait to simplify encoder representation elsewhere
+pub trait HandshakeEncoderMarker:
+    AddressEncoder<Address = Address, Error = EncodingError, RecipientIdentifier = LockingScript>
+{
+}
+
+/// The standard encoder for Bitcoin networks. Parameterized by a `NetworkParams` type and an
+/// `coins_bip32::Encoder`. It exposes
+#[derive(Debug, Clone, PartialEq)]
+pub struct HandshakeEncoder<P: NetworkParams>(PhantomData<fn(P) -> P>);
+
+impl<P: NetworkParams> AddressEncoder for HandshakeEncoder<P> {
+    type Address = Address;
+    type Error = EncodingError;
+    type RecipientIdentifier = LockingScript;
+
+    fn encode_address(s: &LockingScript) -> EncodingResult<Address> {
+        let mut data = Vec::with_capacity(s.serialized_length());
+        s.write_to(&mut data).unwrap();
+
+        match s.standard_type() {
+            LockingScriptType::WSH(_) => Ok(Address::WSH(encode_bech32(P::HRP, &data)?)),
+            LockingScriptType::WPKH(_) => Ok(Address::WPKH(encode_bech32(P::HRP, &data)?)),
+            LockingScriptType::OP_RETURN(_) => {
+                Ok(Address::OP_RETURN(encode_bech32(P::HRP, &data)?))
+            }
+            LockingScriptType::NonStandard => Err(EncodingError::UnknownScriptType),
+        }
+    }
+
+    fn decode_address(addr: &Address) -> EncodingResult<LockingScript> {
+        match &addr {
+            Address::WPKH(s) | Address::WSH(s) | Address::OP_RETURN(s) => {
+                decode_bech32(P::HRP, &s).map(|v| v.into())
+            }
+        }
+    }
+
+    fn string_to_address(string: &str) -> EncodingResult<Address> {
+        let s = string.to_owned();
+        let result = decode_bech32(P::HRP, &s)?;
+
+        let (version_and_len, data) = result.split_at(2);
+        let version = version_and_len[0];
+        let len = version_and_len[1];
+
+        if version == 31 {
+            return Ok(Address::OP_RETURN(s));
+        }
+
+        if len as usize != data.len() {
+            return Err(EncodingError::InvalidSizeError);
+        }
+
+        // Only segwit version 0 is currently defined.
+        if version == 0 {
+            match len {
+                20 => return Ok(Address::WPKH(s)),
+                32 => return Ok(Address::WSH(s)),
+                _ => return Err(EncodingError::UnknownScriptType),
```

Comment:
This should be a wrong script length error as anything that is version 0 with a length different than 20 or 32 is invalid by consensus rules. See: https://github.com/handshake-org/hsd/blob/fb6ce9fe4a884048e54654a591f0b002e829b81f/lib/script/script.js#L2282